### PR TITLE
First release PR -- reviews wanted :)

### PR DESCRIPTION
**Pull request by glormph**
Many thanks to contributing to nf-core/ddamsproteomics!

Please fill in the appropriate checklist below (delete whatever is not relevant). These are the most common things requested on pull requests (PRs).

## PR checklist
 - [ ] This comment contains a description of changes (with reason)
 - [ ] If you've fixed a bug or added code that should be tested, add tests!
 - [ ] If necessary, also make a PR on the [nf-core/ddamsproteomics branch on the nf-core/test-datasets repo]( https://github.com/nf-core/test-datasets/pull/new/nf-core/ddamsproteomics)
 - [ ] Ensure the test suite passes (`nextflow run . -profile test,docker`).
 - [X] Make sure your code lints (`nf-core lint .`).
 - [X] Documentation in `docs` is updated
 - [ ] `CHANGELOG.md` is updated
 - [X] `README.md` is updated

**Learn more about contributing:** https://github.com/nf-core/ddamsproteomics/tree/master/.github/CONTRIBUTING.md

First release PR. The whole pipeline lints and tests run fine *on the dev branch*.

I havent gotten my head around how the combination of `bump-version` and linting decides about the container definition, but `bump-version` did not accept `nfcore/ddamsproteomics:dev` which had to be changed to `latest` before it actually bumped it to 1.0.0. Hence travis will not checkout since the `latest` container is based off this repo's `master` branch and that is behind.

But now I'd like to learn from the mistakes I've made, so bring on the reviews. :)

### CODE REVIEW

**Comment by likelet** [comment]
I think you may have a logo first

File: main.nf [hunk 1]
```diff
@@ -56,30 +78,66 @@ if (params.help){
...
+  trainingpep = file(params.pipep)
+  if( !trainingpep.exists() ) exit 1, "Peptide pI data file not found: ${params.pipep}"
+}
 output_docs = file("$baseDir/docs/output.md")
 
-// Validate inputs
-if ( params.fasta ){
-    fasta = file(params.fasta)
-    if( !fasta.exists() ) exit 1, "Fasta file not found: ${params.fasta}"
+// Parse input variables values
```
Comment (apeltzer) [inline main.nf]: This needs more explicit details: What is happening here? Is this a set of defaults but can be set by users?

File: main.nf [hunk 2]
```diff
@@ -56,30 +78,66 @@ if (params.help){
...
+accolmap = [peptides: 12, proteins: 14, genes: 17, assoc: 18]
+setdenoms = [:]
```
Comment (apeltzer) [inline main.nf:124]: Same here...

File: main.nf [hunk 3]
```diff
@@ -56,30 +78,66 @@ if (params.help){
...
+if (!(params.noquant) && params.isobaric) {
+  params.denoms.tokenize(' ').each{ it -> x=it.tokenize(':'); setdenoms.put(x[0], x[1..-1])}
 }
+normalize = (params.normalize && params.isobaric) ? true: false
+activations = [hcd:'High-energy collision-induced dissociation', cid:'Collision-induced dissociation', etd:'Electron transfer dissociation']
+activationtype = activations[params.activation]
+plextype = params.isobaric ? params.isobaric.replaceFirst(/[0-9]+plex/, "") : 'false'
+massshift = [tmt:0.0013, itraq:0.00125, false:0][plextype]
```
Comment (apeltzer) [inline main.nf]: Setting these values as defaults or can they be set by users?

File: main.nf [hunk 4]
```diff
@@ -128,15 +162,28 @@ log.info """=======================================================
...
+summary['Output genes'] = params.genes
+summary['Output symbols'] = params.symbols
+summary['Custom FASTA delimiter'] = params.fastadelim 
+summary['Custom FASTA gene field'] = params.genefield
+summary['Premade quant data SQLite'] = params.quantlookup
+summary['Fractionated sample'] = fractionation
+summary['HiRIEF'] = params.hirief 
+summary['peptide pI data'] = params.pipep
+summary['Only output peptides'] = params.onlypeptides
+summary['Do not quantify'] = params.noquant
```
Comment (apeltzer) [inline main.nf:183]: Nice!

File: main.nf [hunk 5]
```diff
@@ -191,67 +238,615 @@ process get_software_versions {
...
+} else {
+  Channel
+    .from(file("${params.mzmldef}").readLines())
+    .map { it -> it.tokenize('\t') }
+    .set { mzml_in }
+}
 
+mzml_in
+  .tap { sets }
+  .map { it -> [file(it[0]), it[1], it[2] ? it[2] : it[1], it[3] ? it[3] : 'NA' ]} // create file, set plate to setname, and fraction to NA if there is none
```
Comment (apeltzer) [inline main.nf]: Need a bit more verbose code documentation here: Hard to understand why you do it this way..!

File: main.nf [hunk 6]
```diff
@@ -191,67 +238,615 @@ process get_software_versions {
...
+
+  script:
+  """
+  msslookup spectra -i ${mzmlfiles.join(' ')} --setnames ${setnames.join(' ')}
+  """
 }
 
 
+isoquant_amount = params.isobaric ? amount_mzml.value : 1
+isobaricxml
```
Comment (apeltzer) [inline main.nf:350]: More documentation please!!!

File: main.nf [hunk 7]
```diff
@@ -191,67 +238,615 @@ process get_software_versions {
...
+  .ifEmpty(['NA', 'NA', 'NA'])
+  .buffer(size: isoquant_amount)
+  .map { it.sort({a, b -> a[0] <=> b[0]}) }
+  .map { it -> [it.collect() { it[0] }, it.collect() { it[1] }] } // samples, isoxml
+  .set { isofiles_sets }
+
+kronik_out
```
Comment (apeltzer) [inline main.nf:358]: More documentation please!!!

**Review by apeltzer — CHANGES_REQUESTED** [review]
Please document the code much more :-)

**Comment by glormph** [comment]
Thanks for the comments @apeltzer! :)

**Comment by glormph** [comment]
I assume I need to do a lot more, but unsure what?

**Comment by apeltzer** [comment]
I'll have a look later :-)

File: nextflow.config [hunk 8]
```diff
@@ -33,6 +31,8 @@ profiles {
   docker {
     docker.enabled = true
     process.container = params.container
+  fixOwnership = true
```
Comment (apeltzer) [inline nextflow.config]: Indentation fix?

File: nextflow.config [hunk 9]
```diff
@@ -33,6 +31,8 @@ profiles {
...
+  runOptions = "-u \$(id -u):\$(id -g)"
```
Comment (apeltzer) [inline nextflow.config]: Same here

File: bin/scrape_software_versions.py [hunk 10]
```diff
@@ -4,16 +4,13 @@
 import re
 
 regexes = {
-    'nf-core/lehtio-quant-proteomics': ['v_pipeline.txt', r"(\S+)"],
+    'nf-core/ddamsproteomics': ['v_pipeline.txt', r"(\S+)"],
```
Comment (apeltzer) [inline bin/scrape_software_versions.py:7]: Maybe add your software dependencies in here as well?

File: main.nf [hunk 11]
```diff
@@ -191,67 +235,655 @@ process get_software_versions {
...
+
+  output:
+  set val(sample), file("${sample}.kr"), file(infile) into kronik_out
+  set val(sample), file("${infile}.consensusXML") optional true into isobaricxml
+
+  script:
+  activationtype = [hcd:'High-energy collision-induced dissociation', cid:'Collision-induced dissociation', etd:'Electron transfer dissociation'][params.activation]
+  massshift = [tmt:0.0013, itraq:0.00125, false:0][plextype]
+  """
+  cp $hkconf config
```
Comment (apeltzer) [inline main.nf]: This looks a bit weird to me: The $hkconf is already staged by Nextflow in the process specific work directory - no need to copy anything here...

If you need the content in a separate file, you could still do a simple `cat $hkconfig > config` here...

File: main.nf [hunk 12]
```diff
@@ -191,67 +235,655 @@ process get_software_versions {
...
+  echo "$infile" hardklor.out >> config
+  hardklor config
+  kronik -c 5 -d 3 -g 1 -m 8000 -n 600 -p 10 hardklor.out ${sample}.kr
```
Comment (apeltzer) [inline main.nf:317]: And maybe explain a little what you're doing here :-)

File: main.nf [hunk 13]
```diff
@@ -191,67 +235,655 @@ process get_software_versions {
...
+  set val(td), file("${td}_psmtable.txt") into psm_result
+  set val(td), file({setnames.collect() { it + '.tsv' }}) into setpsmtables
+  set val(td), file("${td}_psmlookup.sql") into psmlookup
+
+  script:
+  """
+  msspsmtable merge -i psms* -o psms.txt
+  msspsmtable conffilt -i psms.txt -o filtpsm --confidence-better lower --confidence-lvl 0.01 --confcolpattern 'PSM q-value'
+  msspsmtable conffilt -i filtpsm -o filtpep --confidence-better lower --confidence-lvl 0.01 --confcolpattern 'peptide q-value'
+  cp lookup psmlookup
```
Comment (apeltzer) [inline main.nf]: `cat lookup > psmlookup` is more explanatory for users checking this code here 👍 

If you don't need the file as an extra one - just use $lookup in the next steps...

File: main.nf [hunk 14]
```diff
@@ -191,67 +235,655 @@ process get_software_versions {
...
+  msspsmtable specdata -i filtpep --dbfile psmlookup -o prepsms.txt
+  ${!params.noquant ? "msspsmtable quant -i prepsms.txt -o qpsms.txt --dbfile psmlookup --precursor ${params.isobaric && td=='target' ? '--isobaric' : ''}" : 'mv prepsms.txt qpsms.txt'}
+  sed 's/\\#SpecFile/SpectraFile/' -i qpsms.txt
+  ${!params.onlypeptides ? "msspsmtable genes -i qpsms.txt -o gpsms --dbfile psmlookup" : ''}
+  ${!params.onlypeptides ? "msslookup proteingroup -i qpsms.txt --dbfile psmlookup" : ''}
+  ${!params.onlypeptides ? "msspsmtable proteingroup -i gpsms -o pgpsms --dbfile psmlookup" : 'mv qpsms.txt pgpsms'}
+  ${params.hirief ? "peptide_pi_annotator.py -i $trainingpep -p pgpsms --o dppsms --stripcolpattern Strip --pepcolpattern Peptide --fraccolpattern Fraction --strippatterns ${allstrips.join(' ')} --intercepts ${allstrips.collect() { params.strips[it].intercept}.join(' ')} --widths ${allstrips.collect() { params.strips[it].fr_width}.join(' ')} --ignoremods \'*\'" : ''}
+  msspsmtable split -i ${params.hirief ? 'dppsms' : 'pgpsms'} --bioset
+  mv ${params.hirief ? 'dppsms' : 'pgpsms'} ${td}_psmtable.txt
+  mv psmlookup ${td}_psmlookup.sql
```
Comment (apeltzer) [inline main.nf]: You could use nextflows file renaming functionality to put this into a channel with a different name, thus not needing this...

File: main.nf [hunk 15]
```diff
@@ -191,67 +235,655 @@ process get_software_versions {
...
+  set val(setname), file('normratiosused') optional true into normratios
+  set val(setname), val(td), file(psms), file('proteins'), val('proteins') into proteins
+  set val(setname), val(td), file(psms), file('genes'), val('genes') into genes
+  set val(setname), val(td), file(psms), file('symbols'), val('assoc') into symbols
+  set val(setname), file('proteinratios') optional true into proteinratios
+
+  script:
+  col = accolmap.peptides + 1  // psm2pep adds a column
+  """
+  msspeptable psm2pep -i psms -o peptides --scorecolpattern svm --spectracol 1 ${!params.noquant && params.isobaric && td == 'target' ? "--isobquantcolpattern plex" : "" } ${!params.noquant ? "--ms1quantcolpattern area" : ""}
```
Comment (apeltzer) [inline main.nf:654]: This needs more comments!!

File: main.nf [hunk 16]
```diff
@@ -191,67 +235,655 @@ process get_software_versions {
...
+  paste <( cut -f ${col} peptides) <( cut -f 1-${col-1},${col+1}-500 peptides) > peptide_table.txt
+  echo Protein accession |tee proteins genes symbols
+  tail -n+2 psms|cut -f ${accolmap.proteins}|grep -v '\\;'|grep -v "^\$"|sort|uniq >> proteins
```
Comment (apeltzer) [inline main.nf:659]: Especially this part - I get what the command does, but *why* you are doing this isn't clear to me :-)

**Review by apeltzer — CHANGES_REQUESTED** [review]
Some comments from my side: Mostly documentation++ ;-)

**Comment by glormph** [comment]
Answered the questions :)

**Comment by glormph** [comment]
Ping this PR. Do I need to do anything specific for the review somewhere on github? Or are commits enough? Due to requests from the lab I'm preparing updates for version 1.1 already :)

**Comment by apeltzer** [comment]
There are some errors in the CI:

```
ERROR: Test Failures:
  http://nf-co.re/errors#5: CI is not pulling the correct docker image. Should be:
    'docker pull nfcore/ddamsproteomics:dev'
  http://nf-co.re/errors#5: CI is not tagging docker image correctly. Should be:
    'docker tag nfcore/ddamsproteomics:dev nfcore/ddamsproteomics:1.0.0'
```

and here:

```
ERROR ~ Error executing process > 'msgfPlus (1)'
Caused by:
  Process `msgfPlus (1)` terminated with an error exit status (1)
Command executed:
  msgf_plus -Xmx16G -d db.fa -s set1_518_scans_QE_tmt10_fasp_cellines_human.mzML -o "set1_518_scans_QE_tmt10_fasp_cellines_human.mzid" -thread 12 -mod tmtmods.txt -tda 0 -t 10.0ppm -ti -1,2 -m 0 -inst 3 -e 1 -protocol 4 -ntt 2 -minLength 7 -maxLength 50 -minCharge 2 -maxCharge 6 -n 1 -addFeatures 1
  msgf_plus -Xmx3500M edu.ucsd.msjava.ui.MzIDToTsv -i "set1_518_scans_QE_tmt10_fasp_cellines_human.mzid" -o out.mzid.tsv
  rm db.c*
Command exit status:
  1
Command output:
  MS-GF+ Release (v2017.07.21) (21 July 2017)
  Loading database files...
  Warning: Sequence database contains 134 counts of letter 'B', which does not correspond to an amino acid.
  Warning: Sequence database contains 116 counts of letter 'U', which does not correspond to an amino acid.
  Warning: Sequence database contains 15124 counts of letter 'X', which does not correspond to an amino acid.
  Warning: Sequence database contains 128 counts of letter 'Z', which does not correspond to an amino acid.
  Creating the suffix array indexed file... Size: 67914521
  AlphabetSize: 28
  Suffix creation: 0.00% complete.
  Suffix creation: 14.72% complete.
  Suffix creation: 29.45% complete.
  Suffix creation: 44.17% complete.
  Suffix creation: 58.90% complete.
  Suffix creation: 73.62% complete.
  Suffix creation: 88.35% complete.
  Sorting 0.00% complete.
  Sorting 5.81% complete.
  Sorting 11.62% complete.
  Sorting 17.43% complete.
  Sorting 23.24% complete.
  Sorting 29.05% complete.
  Sorting 34.86% complete.
  Sorting 40.67% complete.
  Sorting 46.48% complete.
  Sorting 52.29% complete.
  Sorting 58.10% complete.
  Sorting 63.91% complete.
  Sorting 69.73% complete.
  Sorting 75.54% complete.
  Sorting 81.35% complete.
  Sorting 87.16% complete.
  Sorting 92.97% complete.
  Sorting 98.78% complete.
  Loading database finished (elapsed time: 65.32 sec)
  Reading spectra...
Command error:
  Exception in thread "main" java.lang.ExceptionInInitializerError
  	at uk.ac.ebi.jmzml.xml.Constants.<clinit>(Constants.java:56)
  	at uk.ac.ebi.jmzml.xml.xxindex.MzMLIndexerFactory$MzMlIndexerImpl.<init>(MzMLIndexerFactory.java:82)
  	at uk.ac.ebi.jmzml.xml.xxindex.MzMLIndexerFactory$MzMlIndexerImpl.<init>(MzMLIndexerFactory.java:52)
  	at uk.ac.ebi.jmzml.xml.xxindex.MzMLIndexerFactory.buildIndex(MzMLIndexerFactory.java:49)
  	at uk.ac.ebi.jmzml.xml.io.MzMLUnmarshaller.<init>(MzMLUnmarshaller.java:121)
  	at uk.ac.ebi.jmzml.xml.io.MzMLUnmarshaller.<init>(MzMLUnmarshaller.java:100)
  	at edu.ucsd.msjava.mzml.MzMLAdapter.<init>(MzMLAdapter.java:27)
  	at edu.ucsd.msjava.msutil.SpectraAccessor.getSpecMap(SpectraAccessor.java:38)
  	at edu.ucsd.msjava.ui.MSGFPlus.runMSGFPlus(MSGFPlus.java:217)
  	at edu.ucsd.msjava.ui.MSGFPlus.runMSGFPlus(MSGFPlus.java:105)
  	at edu.ucsd.msjava.ui.MSGFPlus.main(MSGFPlus.java:56)
  Caused by: javax.xml.bind.DataBindingException: javax.xml.bind.JAXBException
   - with linked exception:
  [java.lang.ClassNotFoundException: com.sun.xml.internal.bind.v2.ContextFactory]
  	at javax.xml.bind.JAXB.unmarshal(JAXB.java:186)
  	at uk.ac.ebi.jmzml.MzMLElement.loadProperties(MzMLElement.java:609)
  	at uk.ac.ebi.jmzml.MzMLElement.<init>(MzMLElement.java:564)
  	at uk.ac.ebi.jmzml.MzMLElement.<clinit>(MzMLElement.java:83)
  	... 11 more
  Caused by: javax.xml.bind.JAXBException
   - with linked exception:
  [java.lang.ClassNotFoundException: com.sun.xml.internal.bind.v2.ContextFactory]
  	at javax.xml.bind.ContextFinder.newInstance(ContextFinder.java:241)
  	at javax.xml.bind.ContextFinder.find(ContextFinder.java:455)
  	at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:652)
  	at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:599)
  	at javax.xml.bind.JAXB$Cache.<init>(JAXB.java:127)
  	at javax.xml.bind.JAXB.getContext(JAXB.java:154)
  	at javax.xml.bind.JAXB.unmarshal(JAXB.java:183)
  	... 14 more
  Caused by: java.lang.ClassNotFoundException: com.sun.xml.internal.bind.v2.ContextFactory
  	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:583)
  	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
  	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
  	at javax.xml.bind.ContextFinder.safeLoadClass(ContextFinder.java:573)
  	at javax.xml.bind.ContextFinder.newInstance(ContextFinder.java:239)
  	... 20 more
Work dir:
  /home/travis/build/nf-core/ddamsproteomics/tests/work/bd/12fa5ba21144f0aa579d9134e16822
```

**Comment by apeltzer** [comment]
I think pinging @ewels @MaxUlysse for some reviews other than just me would be good 👍

**Comment by glormph** [comment]
Whoa I didnt know you were the only one being pinged, sorry. Will ask some questions in slack.

**Comment by apeltzer** [comment]
No worries Jorrit, I don't know whether the others know this PR would be nice for review 👍 So just thought to link them in here at some point now :-)

**Comment by ewels** [comment]
I prefer to do a release PR on a pseudo-PR against the first commit, to see all of the code instead of just the changes in this review.

If you're done here @apeltzer then I suggest we merge into master and then set up a PR against the first commit as we've done in the other release-ready pipelines recently..

**Comment by ewels** [comment]
...but I guess we're not done here yet as there are quite a few outstanding comments and tests are failing. So only when this is all resolved :)

**Comment by apeltzer** [comment]
Yeah, some issues are still left and/or non-addressed, once this is done I guess we can proceed to do a full review prior to the first release 👍

**Review by apeltzer — APPROVED** [review]

**Comment by apeltzer** [comment]
Please wait until the last test runs through, then feel free to merge it to master and open the pseudo PR to have others also check what's going on in the pipeline. 

I will also go through things then again 👍

**Pull request merged**